### PR TITLE
Utilize Tavily Image Results for Component Creation (Fix for #23)

### PR DESCRIPTION
**Pull request by ms4n**
## Summary

This PR addresses issue #23 by implementing the use of images returned by Tavily's search API in our component creation process, replacing the previous method of using generic placeholder images. It also introduces image descriptions in the PRD tool to enhance the relevance of queried images and adds configuration for S3 upload.

## Changes

1. **Search Service Enhancement**
   - Updated the `performSearch` function to handle and store image results from Tavily.
   - Implemented logic for separate image queries when provided.

2. **PRD Tool Update**
   - Added a new field for image descriptions in the PRD tool input schema.
   - Modified the PRD generation process to include these image descriptions.

3. **Onboarding Tool Controller Update**
   - Updated the product manager tool to incorporate image descriptions and Tavily image results into the PRD.
   - Implemented new search functionality using image descriptions to query Tavily for relevant images.

4. **CTO Tool Controller Improvements**
   - Enhanced the search tool handling to process and include Tavily image results in the response.

5. **Code Service Refinement**
   - Updated the `extractPlaceholderImages` function to parse and return image information from the PRD, including new image descriptions.

6. **Environment Configuration**
   - Added a new environment variable `UPLOAD_TO_S3` to control whether websites are uploaded to S3 or stored locally.

## Impact

- Resolves issue #23 by replacing generic placeholder images with relevant images from Tavily search results.
- Image queries are more targeted and relevant due to specific image descriptions.
- Improved visual relevance and coherence in generated websites.
- More efficient use of available data from our search API partner.
- Flexible storage options for generated websites (S3 or local filesystem).

## Implementation Details

- The changes primarily affect the following files:
  - `server/services/searchService.js`
  - `server/controllers/onboardingToolController.js`
  - `server/controllers/ctoToolController.js`
  - `server/services/codeService.js`
- New environment variable added to `.env.template`:
  ```
  UPLOAD_TO_S3=false
  ```

Attaching screenshots as reference in the comments below

### CODE REVIEW

File: server/controllers/ctoToolController.js [hunk 1]
```diff
@@ -11,39 +15,109 @@ async function handleCTOToolUse({
...
+
+    // Extract images from search results and remove trailing slashes
+    const images = (searchResults.images || []).map((url) =>
+      url.replace(/\/$/, "")
+    );
+
+    let messageContent = [
+      {
+        type: "text",
+        text: `Here are relevant images found for the query "${searchQuery}". Use these images as inspiration or placeholders when designing and creating components for the website:`,
```
Comment (ptmaroct) [inline server/controllers/ctoToolController.js:29]: use the images as is, dont say inspiration or placeholders

File: server/controllers/ctoToolController.js [hunk 2]
```diff
@@ -11,39 +15,109 @@ async function handleCTOToolUse({
...
+        try {
+          // Fetch the image
+          const response = await axios.get(imageUrl, {
+            responseType: "arraybuffer",
+          });
+
+          // Get the Content-Type from the response headers
+          const contentType = response.headers["content-type"];
+
+          // Only process if it's an image
```
Comment (ptmaroct) [inline server/controllers/ctoToolController.js:45]: later: refactor these to methods

File: server/controllers/ctoToolController.js [hunk 3]
```diff
@@ -11,39 +15,109 @@ async function handleCTOToolUse({
...
     ];
-  }
-  else if (tool.name === TOOLS.FILE_CREATOR) {
+  } else if (tool.name === TOOLS.IMAGE_ANALYSIS) {
+    const { image_url, analysis_prompt } = tool.input;
+
+    const analysisResults = await analyzeImage(image_url, analysis_prompt);
+
+    // Process the analysis results
+    const processedResults = {
```
Comment (ptmaroct) [inline server/controllers/ctoToolController.js:101]: dont go with this, let ai decide

File: server/services/codeService.js [hunk 4]
```diff
@@ -1,11 +1,41 @@
...
+} = require("../config/tools");
 const { saveFile } = require("./fileService");
 const { saveFileToS3 } = require("../services/s3Service");
 require("dotenv").config();
 
+const { readFile } = require("./fileService");
+
+async function extractPlaceholderImages(projectFolderName) {
+  const prdContent = await readFile(`${projectFolderName}/readme.md`);
+  const placeholderSection = prdContent.split("Placeholder Images:")[1];
```
Comment (ptmaroct) [inline server/services/codeService.js:14]: improve this and move 1 to some variable with descriptive name

File: server/services/codeService.js [hunk 5]
```diff
@@ -1,11 +1,41 @@
...
+  if (placeholderSection) {
+    return placeholderSection
+      .trim()
+      .split("\n")
+      .map((line) => {
+        const [title, url] = line.split(": ");
```
Comment (ptmaroct) [inline server/services/codeService.js:20]: you can have a json file and fetch values ijjat se

File: server/services/codeService.js [hunk 6]
```diff
@@ -209,24 +240,36 @@ async function codeAssitant({ query, filePath, client }) {
...
+        { role: "user", content: [{ type: "text", text: updatedQuery }] },
+      ],
     });
     const resp = msg.content.find((content) => content.type === "tool_use");
-    console.log("Received response from Anthropic API:", resp);
 
     const { code, description } = resp.input;
 
-    console.log("recieved code", code);
+    console.log("recieved code");
```
Comment (ptmaroct) [inline server/services/codeService.js:251]: remove all or none

**Review by ptmaroct — CHANGES_REQUESTED** [review]

File: server/controllers/ctoToolController.js (hunk 3 above)
Comment (ms4n) [reply server/controllers/ctoToolController.js:101]: I will work on this while building analyze and reapir tool, need to decide on the structure of the analysis and how we can use it

File: server/services/s3Service.js [hunk 7]
```diff
@@ -163,32 +163,36 @@ async function getProjectDirectoryStructure(s3Path) {
...
     // Process directories
     for (const prefix of data.CommonPrefixes || []) {
-      const dirName = prefix.Prefix.split('/').slice(-2)[0];
-      const dirPath = prefix.Prefix.replace('websites/', '');
+      const dirName = prefix.Prefix.split("/").slice(-2)[0];
+      const dirPath = prefix.Prefix.replace("websites/", "");
       structure.push({
         name: dirName,
-        type: 'directory',
+        type: "directory",
```
[the remark this answers is not in the document]
Comment (ptmaroct) [reply server/services/s3Service.js:170]: do it later

**Review by ptmaroct — APPROVED** [review]

**Pull request merged**